### ISSUE TITLE
Skip type annotations from validation of ModifierOrderCheck

### ISSUE THREAD

**Issue by SerialVelocity** [checkstyle/checkstyle#903]
Hey,

Currently, ModifierOrder enforces that annotations must precede all non-annotation modifiers. With Java 8 annotations are allowed in places such as:
```java
List<@Nullable Long> timestamps;
```

This means that combining the above with this:
```java
@Nullable
public Long getStartTimeOrNull() { ... }
```
starts to look a bit odd.

It would be great if it could look like:
```java
public @Nullable Long getStartTimeOrNull() { ... }
```

Would it be possible to add an option to the checkstyle rule to allow us to still check for inconsistent non-annotation modifiers?

I can see a few ways in which this could be made configurable:
 - Ignore all annotations
 - Ignore annotations which match a regex
 - Allow all annotations to be at the beginning or just before the type
 - Allow certain annotations (regex) to be just before the type

I also thought about allowing customisation of the order of modifiers but it doesn't really seem like this would work well since we would probably want annotations which are not relevant to the type to be at the beginning.

Cheers

**Comment by romani** [comment]
@SerialVelocity ,

please update your examples to have whole file content(with imports) to be exact and be compilable. And please share a configuration that you use and error message from Checkstyle. 

We need to see what kind of annotation you try to use:

That annotation is for types:
https://code.google.com/p/jsr-305/source/browse/trunk/ri/src/main/java/javax/annotation/Nullable.java?r=24
http://grepcode.com/file/repo1.maven.org/maven2/edu.washington.cs.types.checker/checker-framework/1.6.4/checkers/nullness/quals/Nullable.java#Nullable

For non types:
http://grepcode.com/file/repo1.maven.org/maven2/com.intellij/annotations/12.0/org/jetbrains/annotations/Nullable.java#Nullable

just for future quick reference: https://blogs.oracle.com/java-platform-group/entry/java_8_s_new_type
https://www.jetbrains.com/idea/documentation/howto.html
http://types.cs.washington.edu/checker-framework/current/checker-framework-manual.html#nullness-checker

**Comment by robtimus** [comment]
I am running into the same issue with a _ThreadSafe_ annotation. When applied to a method it should indicate that the method is thread safe, when applied to a return type it should indicate that the return type is thread safe. Technically it should be possible to annotate both using the same annotation, but javac doesn't allow that yet as it cannot make a distinction between the method annotation and return type annotation.

An example where I am getting CheckStyle warnings:
```java
public @ThreadSafe OutputStream getOutputStream()
```

In this case the annotation shouldn't indicate that the method is thread safe (although it is, but the entire class is thread safe so no need to repeat the annotation on the method), but the returned OutputStream.

**Comment by robtimus** [comment]
Note that CheckStyle gives no warning if the return type annotation is preceded by a generic type:
```java
public static <K, V> @ThreadSafe Map<K, V> createMap(...)
```

**Comment by romani** [comment]
http://checkstyle.sourceforge.net/config_modifier.html#ModifierOrder have nothing to do with annotations, it should care about modifiers only,  so all annotations should be skipped completely.

**Comment by MEZk** [comment]
@romani 
We have ITs for ModifierOrder which check the order of modifiers including annotations order. If we change the behaviour of the check we will break ITs. Should we implement new option?

**Comment by romani** [comment]
@MEZk , please share links to such cases, most likely they are artificial use cases.

ITs are created base on Guava code, do we have such cases in guava ?

**Comment by MEZk** [comment]
@romani 
IT
https://github.com/checkstyle/checkstyle/blob/master/src/it/java/com/google/checkstyle/test/chapter4formatting/rule487modifiers/ModifierOrderTest.java

Input
https://github.com/checkstyle/checkstyle/blob/master/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule487modifiers/InputModifierOrder.java#L24

**Comment by robtimus** [comment]
Line 24 should still be illegal, because annotations cannot be applied to void, only to non-void return types.

**Comment by romani** [comment]
> Line 24 should still be illegal, because annotations cannot be applied to void, only to non-void return types.

If that is compilable code, and all our ITs a compilable. That is valid code. It might be useless, .... , but we should support it. All we need to figure out is make sure that we treat it in the same way it Java do. If that is still annotation on type for Java - we skip it. If that is annotation of the method for Java - we report violation by ModifierOrder. That nuance is reasonable to highlight in javadoc/xdoc.

@MEZk ,

ITs were created during Google style coverage, it was a java7 time. We mostly base them on guava code, we was not aware of such feature.

annotations like on return types become very popular in java8 - https://blogs.oracle.com/java-platform-group/entry/java_8_s_new_type

incorrect violations that we do - http://checkstyle.sourceforge.net/reports/google-style/guava/xref/com/google/common/collect/LinkedListMultimap.html#L684

So after your fix , Guava report need to be regenerated. ITs also has to be adjusted.

**Comment by robtimus** [comment]
The code is valid, but I think that CheckStyle should report a warning about it. This would be a clear example of the annotation being a method annotation, not a return type annotation (since void cannot be annotated). It appears that the distinction between method annotations and return type annotations are sometimes vague, especially with package-private instance methods (where no other modifiers exist).

**Comment by romani** [comment]
fix is merged.

**Comment by pniederw** [comment]
With the latest snapshot (`6.16-20160217.143809-53`), I'm still getting `'@Nullable' annotation modifier does not precede non-annotation modifiers.` in the following two cases:

```java
import javax.annotation.Nullable;

public interface XXX {
    default @Nullable String foo() {
        return null;
    }
}
```

```java
import javax.annotation.Nullable;

public class YYY implements ZZZ {
    @Override
    public @Nullable String foo(Object obj) { ... }
}
```

Is this a bug?

**Comment by romani** [comment]
yes, issue is reopened.

@MEZk , please take a look.

**Comment by romani** [comment]
fix is merged.
